File: golang/search_binary.go
package main

import "fmt"

func search_binary(a []int, target int) int {
    b := 0
    e := len(a) - 1

    for ; b <= e; {
        m := (b + e) / 2
        if a[m] == target {
            return m
        } else if a[m] < target {
            b = m + 1
        } else {
            e = m - 1
        }
    }

    return -1
}

func main() {
    a := []int{3, 4, 7, 5, 1, 6, 8, 2, 9, 0}
    target := 1
    fmt.Println("Array Elements:", a)
    res := search_binary(a, target)

    if res != -1 {
        fmt.Println("Element found at index:", res)
    } else {
        fmt.Println("Element not found")
    }
}

File: c/search_binary.c
#include <stdio.h>

int search_binary(int *a, int length, int key) {
    int b = 0;
    int e = length - 1;

    while (b <= e) {
        int m = (b + e) / 2;

        if (a[m] < key) {
            b = m + 1;
        } else if (a[m] > key) {
            e = m - 1;
        } else {
            return m;
        }
    }

    return -1;
}

int main() {
    int a [] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 0};
    int length = sizeof(a)/sizeof(a[0]);
    int key = 8;
    int res = search_binary(a, length, key);

    if (res == -1) {
        printf("Element not found\n");
    } else {
        printf("Element found at idx: %d\n", res);
    }

    return 0;
}
